### PR TITLE
feat(portex): add ConnectedFields to represent combined fields

### PR DESCRIPTION
**Pull request by Lee-000**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2g5p6va 增加 ConnectedFields 表示拼接后的 fields，作为统一操作 fields 的入口](https://app.clickup.com/t/2g5p6va)

File: graviti/portex/field.py [hunk 1]
```diff
@@ -85,6 +87,190 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+
+    def __len__(self) -> int:
+        return self._indexes[-1]
+
+    def __getitem__(self, index: Union[int, str]) -> PortexType:
+        if isinstance(index, str):
+            return self._fields[self._keys[index]][index]
+
+        index = self._make_index_nonnegative(index)
+        print(index)
```
Comment (linjiX) [inline graviti/portex/field.py:163]: remove the debug code

File: graviti/portex/field.py [hunk 2]
```diff
@@ -85,6 +87,190 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+            index: The index where the changes happened.
+            length: The length change of the index.
+
+        """
+        for i in range(index + 1, len(self._indexes)):
+            self._indexes[i] += length
+        print("updated indexes", self._indexes)
+
+
+class ConnectedFields:
```
Comment (linjiX) [inline graviti/portex/field.py:132]: why not set the base class of ConnectedFields to `NameOrderedDict` or `MutableMapping`?

File: graviti/portex/field.py [hunk 3]
```diff
@@ -85,6 +87,190 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+        print("insert i, j", i, j)
+
+        fields = self._fields[i]
+        if not isinstance(fields, MutableFields) and j == 0:
+            i = i - 1
+            fields = self._fields[i]
+            j = len(fields)
+
+        try:
+            fields.insert(j, name, portex_type)  # type: ignore[attr-defined]
```
Comment (linjiX) [inline graviti/portex/field.py:231]: what if the input name already exist?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/field.py [hunk 4]
```diff
@@ -85,6 +97,147 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+
+    Raises:
+        ValueError: When there as repeated field names.
+
+    Arguments:
+        fields: The ImmutableFields and MutableFields.
+
+    """
+
+    def __init__(self, fields: Iterable[Union[ImmutableFields, MutableFields]]) -> None:
```
Comment (linjiX) [inline graviti/portex/field.py:111]: do we need a base class of ImmutableFields and MutableFields

File: graviti/portex/field.py [hunk 5]
```diff
@@ -85,6 +97,147 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+        keys = {}
+        indexes = []
```
Comment (linjiX) [inline graviti/portex/field.py:113]: we use `indices` in this repo

File: graviti/portex/field.py [hunk 6]
```diff
@@ -85,6 +97,147 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+        for index, field in enumerate(fields):
+            for sub_index, key in enumerate(field):
+                if key in keys:
+                    raise ValueError(f"Repeated field names '{key}'")
+                keys[key] = index
+                indexes.append((index, sub_index, key))
+
+        self._keys = keys
+        self._indexes = indexes
+        self._fields = list(fields)
```
Comment (linjiX) [inline graviti/portex/field.py:123]: transfer `fields` to `list` before iter it.

File: graviti/portex/field.py [hunk 7]
```diff
@@ -85,6 +97,147 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+
+    def __len__(self) -> int:
+        return len(self._indexes)
+
+    def __getitem__(self, index: Union[int, str]) -> PortexType:
+        if isinstance(index, str):
+            i = self._keys[index]
+            key = index
+        else:
+            i, _, key = self._indexes[self._make_index_nonnegative(index)]
```
Comment (linjiX) [inline graviti/portex/field.py:133]: why we need to make index nonnegative here?

File: graviti/portex/field.py [hunk 8]
```diff
@@ -85,6 +97,147 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+        return self._fields[i][key]
+
+    def __setitem__(self, key: Union[int, str], value: PortexType) -> None:
+        if isinstance(key, int):
+            i, _, key = self._indexes[self._make_index_nonnegative(key)]
+        else:
+            i = -1 if key not in self._keys else self._keys[key]
```
Comment (linjiX) [inline graviti/portex/field.py:140]: `dict.get` ?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/field.py (hunk 4 above)
Comment (Lee-000) [reply graviti/portex/field.py:111]: I don't think it's necessary currently.
And the `ImmutableFields` could be the base class.

File: graviti/portex/field.py [hunk 9]
```diff
@@ -85,6 +98,134 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+
+    Arguments:
+        fields: The ImmutableFields and MutableFields.
+
+    """
+
+    def __init__(self, fields: Iterable[Union[ImmutableFields, MutableFields]]) -> None:
+        fields = list(fields)
+        keys = set()
+        for key in chain(*fields):
```
Comment (linjiX) [inline graviti/portex/field.py:115]: use set to check the repeated names?

File: graviti/portex/field.py [hunk 10]
```diff
@@ -85,6 +98,134 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+    def __iter__(self) -> Iterator[str]:
+        for fields in self._fields:
+            yield from fields
+
+    @property
+    def _keys(self) -> Dict[str, int]:
+        return {key: index for index, field in enumerate(self._fields) for key in field}
+
+    @property
+    def _indices(self) -> List[Tuple[int, int]]:
```
Comment (linjiX) [inline graviti/portex/field.py:158]: use method is better.

File: graviti/portex/field.py [hunk 11]
```diff
@@ -85,6 +98,134 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+        self._fields = fields
+
+    def __len__(self) -> int:
+        return sum(map(len, self._fields))
+
+    def __getitem__(self, index: Union[int, str]) -> PortexType:
+        if isinstance(index, int):
+            i, index = self._indices[index]
+        else:
+            i = self._keys[index]
```
Comment (linjiX) [inline graviti/portex/field.py:129]: use python `ChainMap` or loop rather than `self._keys`?

File: graviti/portex/field.py [hunk 12]
```diff
@@ -85,6 +98,134 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+        if isinstance(key, int):
+            i, key = self._indices[key]
+        else:
+            i = self._keys[key]
+
+        del self._fields[i][key]  # type: ignore[attr-defined]
+
+    def __iter__(self) -> Iterator[str]:
+        for fields in self._fields:
+            yield from fields
```
Comment (linjiX) [inline graviti/portex/field.py:151]: `yield from chain(*self._fields)`?

File: graviti/portex/field.py [hunk 13]
```diff
@@ -85,6 +98,134 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+        Arguments:
+            old_name: The current name of the field to be renamed.
+            new_name: The new name of the field to assign.
+
+        Raises:
+            TypeError: When trying to rename a field in ImmutableFields.
+
+        """
+        fields = self._fields[self._keys[old_name]]
+        if not isinstance(fields, MutableFields):
```
Comment (linjiX) [inline graviti/portex/field.py:223]: do we need to check the type of fields?
can we just call the method and catch the exception?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/field.py [hunk 14]
```diff
@@ -85,6 +100,144 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+            TypeError: When trying to insert a field into ImmutableFields.
+
+        """
+        if name in ChainMap(*self._fields):
+            raise ValueError(f"The '{name}' field already exists")
+
+        i, j = self._get_indices(index)
+
+        fields = self._fields[i]
+        if not isinstance(fields, MutableFields) and j == 0:
```
Comment (linjiX) [inline graviti/portex/field.py:198]: can we remove the `isinstance` check?

File: graviti/portex/field.py [hunk 15]
```diff
@@ -85,6 +100,144 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+        else:
+            fields = self._get_fields(key)
+
+        del fields[key]  # type: ignore[attr-defined]
+
+    def __iter__(self) -> Iterator[str]:
+        yield from chain(*self._fields)
+
+    def _get_fields(
+        self, key: str, raise_error: bool = True
```
Comment (linjiX) [inline graviti/portex/field.py:158]: can we remove the `raise_error` flag?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/field.py [hunk 16]
```diff
@@ -85,6 +100,152 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+    Raises:
+        ValueError: When there as repeated field names.
+
+    Arguments:
+        fields: The ImmutableFields and MutableFields.
+
+    """
+
+    def __init__(self, multi_fields: Iterable[UnionFields]) -> None:
+        multi_fields = list(multi_fields)
```
Comment (linjiX) [inline graviti/portex/field.py:118]: put `multi_fields` directly to `ChainMap`?

File: graviti/portex/field.py [hunk 17]
```diff
@@ -85,6 +100,152 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+        field_names: Set[str] = set()
+        for fields in multi_fields:
+            names = set(fields.keys())
+            repeated_keys = field_names & names
+            if repeated_keys:
+                raise ValueError(f"Repeated field names '{repeated_keys}'")
+            field_names = field_names.union(names)
+
+        chained_fields = ChainMap(*multi_fields)
+        self._chained_fields: MutableMapping[str, PortexType] = chained_fields
```
Comment (linjiX) [inline graviti/portex/field.py:128]: can we use `typing.ChainMap`?

File: graviti/portex/field.py [hunk 18]
```diff
@@ -85,6 +100,152 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+
+        return self._chained_fields[key]
+
+    def __setitem__(self, key: Union[int, str], value: PortexType) -> None:
+        if isinstance(key, int):
+            i, key = self._get_indices(key)
+            fields = self._multi_fields[i]
+        else:
+            fields = (
+                self._get_fields(key) if key in self._chained_fields else self._multi_fields[-1]
```
Comment (linjiX) [inline graviti/portex/field.py:147]: will it be better to use try catch to replace if?

File: graviti/portex/field.py [hunk 19]
```diff
@@ -85,6 +100,152 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+            )
+
+        fields[key] = value  # type: ignore[index]
```
Comment (linjiX) [inline graviti/portex/field.py:150]: if fields is immutable? this line will raise `AttributeError`? user may not know what happens?

File: graviti/portex/field.py [hunk 20]
```diff
@@ -85,6 +100,152 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+
+        i, j = self._get_indices(index)
+
+        try:
+            self._multi_fields[i].insert(j, name, portex_type)  # type: ignore[union-attr]
+            return
+        except AttributeError:
+            try:
+                if j == 0:
+                    fields = self._multi_fields[i - 1]
```
Comment (linjiX) [inline graviti/portex/field.py:208]: what if `i == 0` and `i - 1 == -1`?

File: graviti/portex/field.py [hunk 21]
```diff
@@ -85,6 +100,152 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+        """
+        if name in self._chained_fields:
+            raise ValueError(f"The '{name}' field already exists")
+
+        i, j = self._get_indices(index)
+
+        try:
+            self._multi_fields[i].insert(j, name, portex_type)  # type: ignore[union-attr]
+            return
+        except AttributeError:
```
Comment (linjiX) [inline graviti/portex/field.py:205]:
```python
except AttributeError:
    pass

try:
    ...
```

File: graviti/portex/field.py [hunk 22]
```diff
@@ -85,6 +100,152 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+            portex_type: The new PortexType of the field to convert to.
+
+        Raises:
+            TypeError: When trying to change a field in ImmutableFields.
+
+        """
+        try:
+            self._get_fields(name).astype(name, portex_type)  # type: ignore[union-attr]
+        except AttributeError:
+            raise TypeError("Cannot change immutable fields") from None
```
Comment (linjiX) [inline graviti/portex/field.py:230]: do we need to put "name" to the error message?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/portex/field.py [hunk 23]
```diff
@@ -85,6 +147,142 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+    def __init__(self, multi_fields: Iterable[UnionFields]) -> None:
+        multi_fields = list(multi_fields)
+
+        field_names: Set[str] = set()
+        for fields in multi_fields:
+            names = set(fields.keys())
+            repeated_keys = field_names & names
+            if repeated_keys:
+                raise ValueError(f"Repeated field names '{repeated_keys}'")
+            field_names = field_names.union(names)
```
Comment (linjiX) [inline graviti/portex/field.py:173]: `field_name.update(names)`?

File: graviti/portex/field.py [hunk 24]
```diff
@@ -85,6 +147,142 @@ def rename(self, old_name: str, new_name: str) -> None:
...
+        try:
+            self._multi_fields[i].insert(j, name, portex_type)
+            return
+        except TypeError:
+            pass
+
+        try:
+            if j == 0 and i > 0:
+                fields = self._multi_fields[i - 1]
+                fields.insert(len(fields), name, portex_type)
```
Comment (linjiX) [inline graviti/portex/field.py:258]: `fields[name] = portex_type`?

**Review by linjiX — APPROVED** [review]

**Pull request merged**